### PR TITLE
[Universal][converter] batch mode support

### PR DESCRIPTION
**Pull request by martint-unity**
---
### Purpose of this PR
We need to add batch mode support for our converters. So that our users can run Unity in Batch Mode and still call the converters.

Fixes this [FB Case](https://fogbugz.unity3d.com/f/cases/1375962)

---
### Testing status
Can be tested by creating a script like this
```csharp
public class TEST_BATCH_MODE : MonoBehaviour
{
    // Start is called before the first frame update
    void Start()
    {
        Converters.RunInBatchMode(Converters.ContainerType.BuiltInToURP,
            new List<Converters.TypeName> {Converters.TypeName.Material},
            Converters.Filter.Inclusive);          
    }
}
```

### CODE REVIEW

File: com.unity.render-pipelines.universal/Editor/Converter/ConverterUtils.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace UnityEditor.Rendering.Universal.Converters
+{
+    public class ConverterUtils
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Editor/Converter/ConverterUtils.cs:8]: This should be `static`
```suggestion
    public static class ConverterUtils
```

**Review by Nzollner — APPROVED** [review]
LGTM

**Comment by BoonKiangTan** [comment]
2D functionality verified on branch:

2022.1.0b1.2464
Revision: trunk 4897ee769494
Built: Sat, 27 Nov 2021 21:58:31 GMT
Package branch: universal/converter/batch-mode-support (c0ba4e6)

- Verified converter API for 2D URP upgrading work as intended
devenv_3B6hl7j3cd
 - Ran Unity in batchmode with converter API, verified same behaviour as regular Render Pipeline Converter upgrade path
 cmd_PPVVqsOtAi
 - Verified regular sprites, tilemaps, sprite shapes and rigged characters have no visual difference after upgrade
Regular Render Pipeline Converter upgrade path:
Unity_FyrV7DB6GI
Batchmode Converter API upgrade path:
Unity_ypi8iNMWNY

Note: Script has to be in "Editor" folder and executed method has to be defined as static

**Review by BoonKiangTan — APPROVED** [review]

File: com.unity.render-pipelines.universal/Editor/Converter/Converters.cs [hunk 2]
```diff
@@ -0,0 +1,226 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace UnityEditor.Rendering.Universal
+{
+    /// <summary>
+    /// Used for Converting in Batch Mode.
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Editor/Converter/Converters.cs:9]: This doesn't actually describe the enum. For referring to the batch mode, use `<seealso>` tag to refer to the actual function.
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags#seealso

File: com.unity.render-pipelines.universal/Editor/Converter/Converters.cs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            }
+        }
+
+        internal static void RunInBatchMode(Type containerName, List<Type> converterList, ConverterFilter converterFilter)
+        {
+            List<RenderPipelineConverter> convertersToBatch = new List<RenderPipelineConverter>();
+            // This is just a temp to deal with the Include and Exclude enum
+            List<RenderPipelineConverter> tempConvertersToBatch = new List<RenderPipelineConverter>();
+
+            Debug.Log($"Batch Mode:\nContainer named: {containerName}");
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Editor/Converter/Converters.cs:124]:
```suggestion
            Debug.Log($"Converter Batch Mode: {containerName}");
```

File: com.unity.render-pipelines.universal/Editor/Converter/Converters.cs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+#endif
+                case ConverterId.ParametricToFreeformLight:
+                    return typeof(ParametricToFreeformLightUpgrader);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// The method that will be run when converting assets in batch mode.
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Editor/Converter/Converters.cs:81]: This reads kinds of weird. Isn't this a method that the user themselves would be calling?

File: com.unity.render-pipelines.universal/Editor/Converter/Converters.cs [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        /// <summary>
+        /// The method that will be run when converting the assets in batch mode.
+        /// </summary>
+        /// <param name="containerName">The name of the container which will be batched.</param>
+        /// <param name="converterList">The list of converters that will be either included or excluded from batching. These converters need to be part of the passed in container for them to run.</param>
+        /// <param name="converterFilter">The enum that decide if the list of converters will be included or excluded when batching.</param>
+        public static void RunInBatchMode(ConverterContainerId containerName, List<ConverterId> converterList, ConverterFilter converterFilter)
+        {
+            Type containerType = GetContainerType(containerName);
+            List<Type> converterTypes = new List<Type>();
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Editor/Converter/Converters.cs:102]: Reduce reallocations by reserving enough space initially for the maximum amount of items that will end up in the List.
```suggestion
            List<Type> converterTypes = new List<Type>(converterList.Count);
```

**Review by pbbastian — CHANGES_REQUESTED** [review]
Mostly looks good, I have a few comments

File: com.unity.render-pipelines.universal/Editor/Converter/Converters.cs [hunk 6]
```diff
@@ -0,0 +1,229 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace UnityEditor.Rendering.Universal
+{
+    /// <summary>
+    /// Enum for filtering the list of converters in batch mode./>
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Editor/Converter/Converters.cs:9]: No need to mention that it's an enum. Also you've got an accidental `/>` at the end

File: com.unity.render-pipelines.universal/Editor/Converter/Converters.cs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    /// See also <seealso cref="Converters.RunInBatchMode(UnityEditor.Rendering.Universal.ConverterContainerId, List{UnityEditor.Rendering.Universal.ConverterId}, UnityEditor.Rendering.Universal.ConverterFilter)"/>.)
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Editor/Converter/Converters.cs:10]: `<seealso>` goes after `</summary>`

**Review by pbbastian — CHANGES_REQUESTED** [review]
A few more changes :D

**Review by pbbastian — APPROVED** [review]

**Review by kennytann — APPROVED** [review]

**Review by aleksandrasdzikia — APPROVED** [review]

**Review by phi-lira — APPROVED** [review]

**Pull request merged**